File: src/Billing/ContestService.php
<?php
declare(strict_types=1);
namespace App\Billing;
use App\Infrastructure\{Database,Outbox};
final class ContestService
{
    public function __construct(private Database $db, private Outbox $outbox, private Wallet $wallet) {}
    /** Create a contest template. */
    public function createTemplate(array $input, string $actor): array
    {
        $name = trim((string)($input['name'] ?? ''));
        $slug = trim((string)($input['slug'] ?? ''));
        if ($name === '' || mb_strlen($name) > 100) throw new BillingError('Название: 1–100 символов.');
        if (!preg_match('/^[a-z0-9-]{2,50}$/D', $slug)) throw new BillingError('Слаг: 2–50 символов a-z, 0-9, -.');
        $prizeType = (string)($input['prize_type'] ?? 'days');
        if (!in_array($prizeType, ['days', 'balance', 'traffic'], true)) throw new BillingError('Некорректный тип приза.');
        $rawValue=(string)($input['prize_value'] ?? '');
        $maximum=match($prizeType){'balance'=>100000000,'days'=>3650,'traffic'=>100000};
        if (!ctype_digit($rawValue) || (int)$rawValue<1 || (int)$rawValue>$maximum) throw new BillingError('Некорректное значение приза.');
        $maxWinners=(int)($input['max_winners'] ?? 1);
        $timesPerDay=(int)($input['times_per_day'] ?? 1);
        $cooldownHours=(int)($input['cooldown_hours'] ?? 24);
        if ($maxWinners<1 || $maxWinners>100000 || $timesPerDay<1 || $timesPerDay>100 || $cooldownHours<0 || $cooldownHours>8760) throw new BillingError('Некорректные ограничения конкурса.');
        $id = Database::id();
        $this->db->transaction(function() use ($id,$name,$slug,$input,$prizeType,$rawValue,$maxWinners,$timesPerDay,$cooldownHours,$actor) {
            $this->db->execute(
                'INSERT INTO contest_templates(id,name,slug,description,prize_type,prize_value,max_winners,attempts_per_user,times_per_day,cooldown_hours,is_enabled,created_at) VALUES(?,?,?,?,?,?,?,?,?,?,1,?)',
                [$id, $name, $slug, $input['description'] ?? null, $prizeType, $rawValue, $maxWinners, 1, $timesPerDay, $cooldownHours, time()]
            );
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'contest.created', $id, time()]);
        });
        return $this->db->one('SELECT * FROM contest_templates WHERE id=?', [$id]);
    }
    public function listTemplates(): array
    {
        return $this->db->all('SELECT * FROM contest_templates ORDER BY created_at DESC');
    }
    /** Start a round for a template. */
    public function startRound(string $templateId, string $actor): array
    {
        $t = $this->db->one('SELECT * FROM contest_templates WHERE id=?', [$templateId]);
        if (!$t) throw new BillingError('Конкурс не найден.');
        $id = Database::id();
        $now = time();
        $this->db->transaction(function() use ($id,$templateId,$now,$actor) {
            $this->db->execute('INSERT INTO contest_rounds(id,template_id,status,starts_at,created_at) VALUES(?,?,?,?,?)', [$id, $templateId, 'running', $now, $now]);
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'contest.round_started', $id, $now]);
        });
        return $this->db->one('SELECT * FROM contest_rounds WHERE id=?', [$id]);
    }
    /** Attempt a contest round. Returns ['won'=>bool,'prize'=>?string]. */
    public function attempt(string $userId, string $roundId): array
    {
        return $this->db->transaction(function () use ($userId, $roundId) {
            $round = $this->db->one('SELECT * FROM contest_rounds WHERE id=?' . $this->db->lock(), [$roundId]);
            if (!$round || $round['status'] !== 'running') throw new BillingError('Раунд не активен.');
            $template = $this->db->one('SELECT * FROM contest_templates WHERE id=?'.$this->db->lock(), [$round['template_id']]);
            if (!$template || (int)$template['is_enabled'] !== 1) throw new BillingError('Конкурс отключён.');
            $maximum=match($template['prize_type']){'balance'=>100000000,'days'=>3650,'traffic'=>100000,default=>0};
            if ($maximum===0 || !ctype_digit((string)$template['prize_value']) || (int)$template['prize_value']<1 || (int)$template['prize_value']>$maximum) throw new BillingError('Приз конкурса не настроен.');
            $existing = $this->db->one('SELECT * FROM contest_attempts WHERE round_id=? AND user_id=?', [$roundId, $userId]);
            if ($existing) return ['won' => (int)$existing['won'] === 1, 'prize' => $existing['prize_value'], 'already' => true];
            $user=$this->db->one('SELECT disabled FROM users WHERE id=?',[$userId]);
            if (!$user || (int)$user['disabled']===1) throw new BillingError('Аккаунт недоступен.');
            $now=time();
            $attemptsToday = (int)($this->db->one('SELECT COUNT(*) AS c FROM contest_attempts a JOIN contest_rounds r ON r.id=a.round_id WHERE a.user_id=? AND r.template_id=? AND a.created_at>?', [$userId,$template['id'],$now-86400])['c'] ?? 0);
            if ($attemptsToday >= (int)$template['times_per_day']) throw new BillingError('Лимит попыток на сегодня исчерпан.');
            $latest=$this->db->one('SELECT MAX(a.created_at) AS at FROM contest_attempts a JOIN contest_rounds r ON r.id=a.round_id WHERE a.user_id=? AND r.template_id=?',[$userId,$template['id']]);
            if ($latest['at']!==null && (int)$latest['at']+(int)$template['cooldown_hours']*3600>$now) throw new BillingError('Дождитесь окончания перерыва между попытками.');
            $sub=null;
            if ($template['prize_type']==='days' || $template['prize_type']==='traffic') {
                $limited=$template['prize_type']==='traffic' ? ' AND traffic_limit_gb>0' : '';
                $sub=$this->db->one("SELECT * FROM subscriptions WHERE user_id=? AND status IN ('active','trial') AND expires_at>?$limited ORDER BY expires_at DESC LIMIT 1".$this->db->lock(),[$userId,$now]);
                if (!$sub) throw new BillingError($template['prize_type']==='traffic'?'Для приза нужен действующий тариф с лимитом трафика.':'Для приза нужна действующая подписка.');
            }
            $winners = (int)($this->db->one('SELECT COUNT(*) AS c FROM contest_attempts WHERE round_id=? AND won=1', [$roundId])['c'] ?? 0);
            $won = $winners < (int)$template['max_winners'] && random_int(1, 100) <= 20;
            $prize = $won ? $template['prize_value'] : null;
            $this->db->execute('INSERT INTO contest_attempts(id,round_id,user_id,won,prize_value,created_at) VALUES(?,?,?,?,?,?)', [Database::id(), $roundId, $userId, (int)$won, $prize, time()]);
            if ($won) {
                $this->award($userId, $roundId, $template, $prize, $sub);
            }
            return ['won' => $won, 'prize' => $prize, 'already' => false];
        });
    }
    private function award(string $userId, string $roundId, array $template, string $prize, ?array $sub): void
    {
        $type = $template['prize_type'];
        $value = (int)$prize;
        if ($type === 'balance' && $value > 0) {
            $this->wallet->credit($userId, $value, 'referral_reward', 'Приз за конкурс: '.$template['name']);
        } elseif ($type === 'days' && $value > 0) {
            $base = max(time(), (int)$sub['expires_at']);
            $this->db->execute("UPDATE subscriptions SET expires_at=?,status='active' WHERE id=?", [$base + $value * 86400, $sub['id']]);
            $this->outbox->enqueue('subscription.extend', 'contest-extend:'.$roundId.':'.$userId.':'.$sub['id'], ['subscription_id' => $sub['id']]);
        } elseif ($type === 'traffic' && $value > 0) {
            $this->db->execute('UPDATE subscriptions SET purchased_traffic_gb=purchased_traffic_gb+?,updated_at=? WHERE id=?',[$value,time(),$sub['id']]);
            $this->outbox->enqueue('subscription.traffic','contest-traffic:'.$roundId.':'.$userId.':'.$sub['id'],['subscription_id'=>$sub['id'],'traffic_gb'=>$value]);
        } else {
            throw new BillingError('Приз конкурса не настроен.');
        }
    }
    public function finishRound(string $roundId, string $actor): void
    {
        $this->db->transaction(function() use ($roundId,$actor) {
            $updated=$this->db->execute("UPDATE contest_rounds SET status='finished',ends_at=? WHERE id=? AND status='running'", [time(), $roundId]);
            if ($updated) $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'contest.round_finished', $roundId, time()]);
        });
    }
}

File: rails_backend/app/services/billing/contest_service.rb
module Billing
  class ContestService
    def attempt(user_id:, round_id:)
      ApplicationRecord.transaction do
        round = ContestRound.lock.find_by(id: round_id)
        raise BillingError, "Раунд не активен." unless round&.status == "running"
        template = ContestTemplate.lock.find_by(id: round.template_id)
        raise BillingError, "Конкурс отключён." unless template && template.is_enabled.to_i == 1
        value = template.prize_value.to_s
        maximum = { "balance" => 100_000_000, "days" => 3650, "traffic" => 100_000 }[template.prize_type]
        raise BillingError, "Приз конкурса не настроен." unless maximum && value.match?(/\A\d+\z/) && value.to_i.between?(1, maximum)
        existing = ContestAttempt.find_by(round_id: round.id, user_id: user_id)
        next({ won: existing.won.to_i == 1, prize: existing.prize_value, already: true }) if existing
        user = User.find_by(id: user_id)
        raise BillingError, "Аккаунт недоступен." unless user && user.disabled.to_i.zero?

        now = Time.now.to_i
        user_attempts = ContestAttempt.joins(:round)
          .where(user_id: user_id, contest_rounds: { template_id: template.id })
        raise BillingError, "Лимит попыток на сегодня исчерпан." if user_attempts.where("contest_attempts.created_at > ?", now - 86_400).count >= template.times_per_day.to_i
        latest = user_attempts.maximum(:created_at)
        if latest && latest.to_i + template.cooldown_hours.to_i * 3600 > now
          raise BillingError, "Дождитесь окончания перерыва между попытками."
        end

        subscription = eligible_subscription(user_id, template)
        winners = ContestAttempt.where(round_id: round.id, won: 1).count
        won = winners < template.max_winners.to_i && SecureRandom.random_number(100).between?(0, 19)
        prize = won ? value : nil
        ContestAttempt.create!(id: Infrastructure::IdGenerator.call, round_id: round.id,
          user_id: user_id, won: won ? 1 : 0, prize_value: prize, created_at: now)
        award(user_id, round, template, value.to_i, subscription) if won
        { won: won, prize: prize, already: false }
      end
    rescue ActiveRecord::RecordNotUnique
      prior = ContestAttempt.find_by(round_id: round_id, user_id: user_id)
      raise BillingError, "Попытка конкурса уже обработана." unless prior
      { won: prior.won.to_i == 1, prize: prior.prize_value, already: true }
    end

    def finish_round(round_id:, actor:)
      ApplicationRecord.transaction do
        round = ContestRound.lock.find_by(id: round_id)
        raise BillingError, "Раунд не найден." unless round
        next false unless round.status == "running"
        now = Time.now.to_i
        round.update!(status: "finished", ends_at: now)
        AuditLog.create!(id: Infrastructure::IdGenerator.call, actor: actor,
          action: "contest.round_finished", subject: round.id, created_at: now)
        true
      end
    end

    private

    def eligible_subscription(user_id, template)
      return unless %w[days traffic].include?(template.prize_type)
      scope = Subscription.where(user_id: user_id, status: %w[active trial]).where("expires_at > ?", Time.now.to_i)
      scope = scope.where("traffic_limit_gb > 0") if template.prize_type == "traffic"
      sub = scope.order(expires_at: :desc).lock.first
      raise BillingError, template.prize_type == "traffic" ? "Для приза нужен действующий тариф с лимитом трафика." : "Для приза нужна действующая подписка." unless sub
      sub
    end

    def award(user_id, round, template, value, subscription)
      case template.prize_type
      when "balance"
        WalletService.new.credit(user_id, value, "referral_reward", "Приз за конкурс: #{template.name}")
      when "days"
        now = Time.now.to_i
        subscription.update!(expires_at: [now, subscription.expires_at.to_i].max + value * 86_400,
          status: "active", lifecycle_status: "active", updated_at: now)
        Infrastructure::OutboxService.new.enqueue("subscription.extend",
          "contest-extend:#{round.id}:#{user_id}:#{subscription.id}", { subscription_id: subscription.id })
      when "traffic"
        subscription.update!(purchased_traffic_gb: subscription.purchased_traffic_gb.to_i + value, updated_at: Time.now.to_i)
        Infrastructure::OutboxService.new.enqueue("subscription.traffic",
          "contest-traffic:#{round.id}:#{user_id}:#{subscription.id}", { subscription_id: subscription.id, traffic_gb: value })
      else
        raise BillingError, "Приз конкурса не настроен."
      end
    end
  end
end
